apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -605,21 +608,27 @@ update_state(Data, _DataWas) ->
 health_check_interval(Opts) ->
     maps:get(health_check_interval, Opts, ?HEALTHCHECK_INTERVAL).
 
-maybe_alarm(connected, _ResId, _Error) ->
+maybe_alarm(connected, _ResId, _Error, _PrevError) ->
     ok;
-maybe_alarm(_Status, <<?TEST_ID_PREFIX, _/binary>>, _Error) ->
+maybe_alarm(_Status, <<?TEST_ID_PREFIX, _/binary>>, _Error, _PrevError) ->
     ok;
-maybe_alarm(_Status, ResId, Error) ->
+maybe_alarm(_Status, ResId, Error, none = _PrevError) ->
     HrError =
         case Error of
             undefined -> <<"Unknown reason">>;
             _Else -> emqx_utils:readable_error_msg(Error)
         end,
-    emqx_alarm:activate(
-        ResId,
-        #{resource_id => ResId, reason => resource_down},
-        <<"resource down: ", HrError/binary>>
-    ).
+    ?SAFE_CALL_WITH_LOG(
+        emqx_alarm:activate(
+            ResId,
+            #{resource_id => ResId, reason => resource_down},
+            <<"resource down: ", HrError/binary>>
+        )
+    ),
+    ?tp(resource_activate_alarm, #{resource_id => ResId});
+%% Assume that alarm is already active
+maybe_alarm(_Status, _ResId, _Error, _PrevError) ->
```

Comment:
You are right, we don't update alarm description when moving from one error state to another (new) error state. 
I think it also should be addressed.. It can be easily changed in `maybe_alarm` by comparing prev and current errors, but we need to update emqx_alarm accordingly.

Reply:
I've changed it to avoid activating alarm only if previous error is the same as the current one. 
For `emqx_alarm`, I've added a separate task: https://emqx.atlassian.net/browse/EMQX-9636, as it's not closely connected with this PR..